File: Toggly.FeatureManagement.Next/nextjs-toggly-core/src/telemetry/usage-batcher.ts
import { hashIdentity, toProtobufTimestamp } from './hash.js'

/** Hard caps aligned with PHP/.NET unique-hash limits. */
export const MAX_UNIQUE_USER_HASHES_PER_FEATURE = 10_000
export const MAX_APPLICATION_UNIQUE_USER_HASHES = 10_000
/** Max distinct feature keys retained in one in-memory batch. */
export const MAX_FEATURES_PER_BATCH = 500

export interface VariantStatsAgg {
  checkCount: number
  requestCount: number
  usedCount: number
  viewedCount: number
}

export interface FeatureUsageAgg {
  variantStats: Map<string, VariantStatsAgg>
  uniqueUsersEnabled: Set<number>
  uniqueUsersDisabled: Set<number>
  uniqueUsersUsed: Set<number>
  uniqueUsersViewed: Set<number>
  uniqueUserHashes: Set<number>
  uniqueViewedUserHashes: Set<number>
}

export interface UsageBatcherOptions {
  appKey: string
  environment: string
  instanceName?: string
  appVersion?: string
  processStartTime?: Date
  maxUniqueHashesPerFeature?: number
  maxApplicationUniqueHashes?: number
  maxFeatures?: number
}

export interface FeatureStatPayload {
  appKey: string
  environment: string
  time: { seconds: number; nanos: number }
  stats: Array<{
    feature: string
    uniqueContextIdentifierEnabledCount: number
    uniqueContextIdentifierDisabledCount: number
    uniqueUsersUsedCount: number
    uniqueUserHashes: number[]
    uniqueViewedUserHashes: number[]
    variantStats: Record<string, VariantStatsAgg>
  }>
  totalUniqueUsers: number
  uniqueUserHashes: number[]
  instanceName?: string
  appVersion?: string
  processStartTime?: { seconds: number; nanos: number }
}

/**
 * Wire payload plus uniqueness hash-set snapshots for soft-fail HTTPS restore.
 */
export interface UsageFlushBundle {
  payload: FeatureStatPayload
  uniqueUsersEnabled: Record<string, number[]>
  uniqueUsersDisabled: Record<string, number[]>
  uniqueUsersUsed: Record<string, number[]>
}

function emptyVariant(): VariantStatsAgg {
  return { checkCount: 0, requestCount: 0, usedCount: 0, viewedCount: 0 }
}

function emptyFeature(): FeatureUsageAgg {
  return {
    variantStats: new Map(),
    uniqueUsersEnabled: new Set(),
    uniqueUsersDisabled: new Set(),
    uniqueUsersUsed: new Set(),
    uniqueUsersViewed: new Set(),
    uniqueUserHashes: new Set(),
    uniqueViewedUserHashes: new Set(),
  }
}

/**
 * In-memory feature usage aggregator. Prefer variantStats over legacy scalars
 * (matches .NET TogglyUsageStatsProvider send shape).
 */
export class UsageBatcher {
  private readonly appKey: string
  private readonly environment: string
  private readonly instanceName?: string
  private readonly appVersion?: string
  private readonly processStartTime: Date
  private readonly maxUniqueHashesPerFeature: number
  private readonly maxApplicationUniqueHashes: number
  private readonly maxFeatures: number
  private perFeature = new Map<string, FeatureUsageAgg>()
  private appUnique = new Set<number>()
  private droppedFeatures = false

  constructor(options: UsageBatcherOptions) {
    this.appKey = options.appKey
    this.environment = options.environment
    this.instanceName = options.instanceName
    this.appVersion = options.appVersion
    this.processStartTime = options.processStartTime ?? new Date()
    this.maxUniqueHashesPerFeature =
      options.maxUniqueHashesPerFeature ?? MAX_UNIQUE_USER_HASHES_PER_FEATURE
    this.maxApplicationUniqueHashes =
      options.maxApplicationUniqueHashes ?? MAX_APPLICATION_UNIQUE_USER_HASHES
    this.maxFeatures = options.maxFeatures ?? MAX_FEATURES_PER_BATCH
  }

  private get(feature: string): FeatureUsageAgg | null {
    let agg = this.perFeature.get(feature)
    if (!agg) {
      if (this.perFeature.size >= this.maxFeatures) {
        this.droppedFeatures = true
        return null
      }
      agg = emptyFeature()
      this.perFeature.set(feature, agg)
    }
    return agg
  }

  private getVariant(agg: FeatureUsageAgg, variant: string): VariantStatsAgg {
    let stats = agg.variantStats.get(variant)
    if (!stats) {
      stats = emptyVariant()
      agg.variantStats.set(variant, stats)
    }
    return stats
  }

  private addHashCapped(set: Set<number>, hash: number, max: number): void {
    if (set.has(hash) || set.size < max) {
      set.add(hash)
    }
  }

  private trackAppUnique(hash: number): void {
    this.addHashCapped(this.appUnique, hash, this.maxApplicationUniqueHashes)
  }

  /**
   * Record a feature evaluation.
   *
   * `checkCount` increments on every call. `requestCount` maps to .NET
   * UniqueRequestEnabled/Disabled when `uniqueRequest` is true.
   */
  recordCheck(
    feature: string,
    enabled: boolean,
    identity?: string,
    variant?: string,
    uniqueRequest = false,
  ): void {
    const agg = this.get(feature)
    if (!agg) return

    const name = variant ?? (enabled ? 'enabled' : 'disabled')
    const stats = this.getVariant(agg, name)
    stats.checkCount += 1
    if (uniqueRequest) {
      stats.requestCount += 1
    }

    if (identity) {
      const hash = hashIdentity(identity)
      this.trackAppUnique(hash)
      if (enabled) {
        this.addHashCapped(agg.uniqueUsersEnabled, hash, this.maxUniqueHashesPerFeature)
      } else {
        this.addHashCapped(agg.uniqueUsersDisabled, hash, this.maxUniqueHashesPerFeature)
      }
    }
  }

  recordUsage(feature: string, identity?: string, variant = 'enabled'): void {
    const agg = this.get(feature)
    if (!agg) return

    this.getVariant(agg, variant).usedCount += 1
    if (identity) {
      const hash = hashIdentity(identity)
      this.trackAppUnique(hash)
      this.addHashCapped(agg.uniqueUsersUsed, hash, this.maxUniqueHashesPerFeature)
      this.addHashCapped(agg.uniqueUserHashes, hash, this.maxUniqueHashesPerFeature)
    }
  }

  recordView(feature: string, identity?: string, variant = 'enabled'): void {
    const agg = this.get(feature)
    if (!agg) return

    this.getVariant(agg, variant).viewedCount += 1
    if (identity) {
      const hash = hashIdentity(identity)
      this.trackAppUnique(hash)
      this.addHashCapped(agg.uniqueUsersViewed, hash, this.maxUniqueHashesPerFeature)
      this.addHashCapped(agg.uniqueViewedUserHashes, hash, this.maxUniqueHashesPerFeature)
    }
  }

  isEmpty(): boolean {
    return this.perFeature.size === 0 && this.appUnique.size === 0
  }

  hitFeatureCap(): boolean {
    return this.droppedFeatures
  }

  featureCount(): number {
    return this.perFeature.size
  }

  buildAndReset(): UsageFlushBundle | null {
    if (this.isEmpty()) {
      return null
    }

    const payload: FeatureStatPayload = {
      appKey: this.appKey,
      environment: this.environment,
      time: toProtobufTimestamp(),
      stats: [],
      totalUniqueUsers: this.appUnique.size,
      uniqueUserHashes: [...this.appUnique],
      processStartTime: toProtobufTimestamp(this.processStartTime),
    }

    if (this.instanceName) {
      payload.instanceName = this.instanceName
    }
    if (this.appVersion) {
      payload.appVersion = this.appVersion
    }

    const uniqueUsersEnabled: Record<string, number[]> = {}
    const uniqueUsersDisabled: Record<string, number[]> = {}
    const uniqueUsersUsed: Record<string, number[]> = {}

    for (const [feature, agg] of this.perFeature) {
      const variantStats: Record<string, VariantStatsAgg> = {}
      for (const [name, stats] of agg.variantStats) {
        if (
          stats.checkCount > 0 ||
          stats.requestCount > 0 ||
          stats.usedCount > 0 ||
          stats.viewedCount > 0
        ) {
          variantStats[name] = { ...stats }
        }
      }

      if (agg.uniqueUsersEnabled.size > 0) {
        uniqueUsersEnabled[feature] = [...agg.uniqueUsersEnabled]
      }
      if (agg.uniqueUsersDisabled.size > 0) {
        uniqueUsersDisabled[feature] = [...agg.uniqueUsersDisabled]
      }
      if (agg.uniqueUsersUsed.size > 0) {
        uniqueUsersUsed[feature] = [...agg.uniqueUsersUsed]
      }

      payload.stats.push({
        feature,
        uniqueContextIdentifierEnabledCount: agg.uniqueUsersEnabled.size,
        uniqueContextIdentifierDisabledCount: agg.uniqueUsersDisabled.size,
        uniqueUsersUsedCount: agg.uniqueUsersUsed.size,
        uniqueUserHashes: [...agg.uniqueUserHashes],
        uniqueViewedUserHashes: [...agg.uniqueViewedUserHashes],
        variantStats,
      })
    }

    this.perFeature = new Map()
    this.appUnique = new Set()
    this.droppedFeatures = false
    return {
      payload,
      uniqueUsersEnabled,
      uniqueUsersDisabled,
      uniqueUsersUsed,
    }
  }

  restoreFromBundle(bundle: UsageFlushBundle): void {
    const { payload } = bundle

    for (const hash of payload.uniqueUserHashes ?? []) {
      this.trackAppUnique(hash)
    }

    for (const stat of payload.stats ?? []) {
      const feature = stat.feature
      if (!feature) continue
      const agg = this.get(feature)
      if (!agg) continue

      for (const [name, vs] of Object.entries(stat.variantStats ?? {})) {
        const target = this.getVariant(agg, name)
        target.checkCount += vs.checkCount ?? 0
        target.requestCount += vs.requestCount ?? 0
        target.usedCount += vs.usedCount ?? 0
        target.viewedCount += vs.viewedCount ?? 0
      }
      for (const hash of stat.uniqueUserHashes ?? []) {
        this.addHashCapped(agg.uniqueUserHashes, hash, this.maxUniqueHashesPerFeature)
      }
      for (const hash of stat.uniqueViewedUserHashes ?? []) {
        this.addHashCapped(agg.uniqueViewedUserHashes, hash, this.maxUniqueHashesPerFeature)
        this.addHashCapped(agg.uniqueUsersViewed, hash, this.maxUniqueHashesPerFeature)
      }
    }

    this.mergeUniqueHashMap(bundle.uniqueUsersEnabled, 'uniqueUsersEnabled')
    this.mergeUniqueHashMap(bundle.uniqueUsersDisabled, 'uniqueUsersDisabled')
    this.mergeUniqueHashMap(bundle.uniqueUsersUsed, 'uniqueUsersUsed')
  }

  private mergeUniqueHashMap(
    snapshot: Record<string, number[]>,
    field: 'uniqueUsersEnabled' | 'uniqueUsersDisabled' | 'uniqueUsersUsed',
  ): void {
    for (const [feature, hashes] of Object.entries(snapshot ?? {})) {
      const agg = this.get(feature)
      if (!agg) continue
      for (const hash of hashes) {
        this.addHashCapped(agg[field], hash, this.maxUniqueHashesPerFeature)
        this.trackAppUnique(hash)
      }
    }
  }
}

File: toggly-rust/toggly/src/telemetry/usage_batcher.rs
//! In-memory feature usage aggregator (`Usage.SendStats` payload shape).

use super::hash::{hash_identity, now_protobuf_timestamp, to_protobuf_timestamp_millis};
use chrono::{DateTime, Utc};
use parking_lot::Mutex;
use std::collections::{HashMap, HashSet};

/// Per-variant counters for a feature.
#[derive(Debug, Clone, Default, PartialEq, Eq)]
pub struct VariantStatsAgg {
    /// Evaluation / check count.
    pub check_count: i32,
    /// Unique-request count (first access in a logical request).
    pub request_count: i32,
    /// Feature "used" / interaction count.
    pub used_count: i32,
    /// Feature "viewed" / rendered count.
    pub viewed_count: i32,
}

impl VariantStatsAgg {
    /// Serialize counters to the Usage wire field names.
    pub fn to_wire(&self) -> VariantStatsWire {
        VariantStatsWire {
            check_count: self.check_count,
            request_count: self.request_count,
            used_count: self.used_count,
            viewed_count: self.viewed_count,
        }
    }

    fn is_empty(&self) -> bool {
        self.check_count == 0
            && self.request_count == 0
            && self.used_count == 0
            && self.viewed_count == 0
    }
}

/// Wire shape for one variantStats entry.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct VariantStatsWire {
    /// checkCount
    pub check_count: i32,
    /// requestCount
    pub request_count: i32,
    /// usedCount
    pub used_count: i32,
    /// viewedCount
    pub viewed_count: i32,
}

#[derive(Debug, Clone, Default)]
struct FeatureUsageAgg {
    variant_stats: HashMap<String, VariantStatsAgg>,
    unique_users_enabled: HashSet<i32>,
    unique_users_disabled: HashSet<i32>,
    unique_users_used: HashSet<i32>,
    unique_user_hashes: HashSet<i32>,
    unique_viewed_user_hashes: HashSet<i32>,
}

/// Per-feature row in a FeatureStat payload.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct StatMessagePayload {
    /// Feature key.
    pub feature: String,
    /// Unique enabled context identifiers.
    pub unique_context_identifier_enabled_count: i32,
    /// Unique disabled context identifiers.
    pub unique_context_identifier_disabled_count: i32,
    /// Unique users who used the feature.
    pub unique_users_used_count: i32,
    /// Unique identity hashes for used tracking.
    pub unique_user_hashes: Vec<i32>,
    /// Unique identity hashes for viewed tracking.
    pub unique_viewed_user_hashes: Vec<i32>,
    /// Multi-variant stats map (preferred over legacy scalars).
    pub variant_stats: HashMap<String, VariantStatsWire>,
}

/// FeatureStat-shaped payload ready for gRPC.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct FeatureStatPayload {
    /// Application key.
    pub app_key: String,
    /// Environment name.
    pub environment: String,
    /// Packet time (`seconds`, `nanos`).
    pub time: (i64, i32),
    /// Per-feature stats.
    pub stats: Vec<StatMessagePayload>,
    /// Application-level unique user count.
    pub total_unique_users: i32,
    /// Application-level unique user hashes.
    pub unique_user_hashes: Vec<i32>,
    /// Optional instance name.
    pub instance_name: Option<String>,
    /// Optional app version.
    pub app_version: Option<String>,
    /// Process start time (`seconds`, `nanos`).
    pub process_start_time: (i64, i32),
}

/// In-memory snapshot of usage aggregates taken before a send attempt.
#[derive(Debug, Clone)]
pub struct UsageSnapshot {
    per_feature: HashMap<String, FeatureUsageAgg>,
    app_unique: HashSet<i32>,
}

/// Accumulate feature usage and build FeatureStat payloads.
///
/// Prefer `variant_stats` over legacy scalars (matches .NET / Node / Ruby).
pub struct UsageBatcher {
    app_key: String,
    environment: String,
    instance_name: Option<String>,
    app_version: Option<String>,
    process_start_time: DateTime<Utc>,
    inner: Mutex<UsageInner>,
}

#[derive(Default)]
struct UsageInner {
    per_feature: HashMap<String, FeatureUsageAgg>,
    app_unique: HashSet<i32>,
}

impl UsageBatcher {
    /// Create an empty usage aggregator for one app/environment.
    pub fn new(
        app_key: impl Into<String>,
        environment: impl Into<String>,
        instance_name: Option<String>,
        app_version: Option<String>,
        process_start_time: Option<DateTime<Utc>>,
    ) -> Self {
        Self {
            app_key: app_key.into(),
            environment: environment.into(),
            instance_name,
            app_version,
            process_start_time: process_start_time.unwrap_or_else(Utc::now),
            inner: Mutex::new(UsageInner::default()),
        }
    }

    /// Record a feature evaluation.
    pub fn record_check(
        &self,
        feature: &str,
        enabled: bool,
        identity: Option<&str>,
        variant: Option<&str>,
        unique_request: bool,
    ) {
        let hashed = identity.filter(|s| !s.is_empty()).map(hash_identity);
        let name = variant
            .map(str::to_string)
            .unwrap_or_else(|| if enabled { "enabled" } else { "disabled" }.to_string());

        let mut guard = self.inner.lock();
        if let Some(h) = hashed {
            guard.app_unique.insert(h);
        }
        let agg = guard.per_feature.entry(feature.to_string()).or_default();
        let stats = agg.variant_stats.entry(name).or_default();
        stats.check_count += 1;
        if unique_request {
            stats.request_count += 1;
        }
        if let Some(h) = hashed {
            if enabled {
                agg.unique_users_enabled.insert(h);
            } else {
                agg.unique_users_disabled.insert(h);
            }
        }
    }

    /// Record a feature used/interaction event.
    pub fn record_usage(&self, feature: &str, identity: Option<&str>, variant: &str) {
        let hashed = identity.filter(|s| !s.is_empty()).map(hash_identity);
        let mut guard = self.inner.lock();
        if let Some(h) = hashed {
            guard.app_unique.insert(h);
        }
        let agg = guard.per_feature.entry(feature.to_string()).or_default();
        agg.variant_stats
            .entry(variant.to_string())
            .or_default()
            .used_count += 1;
        if let Some(h) = hashed {
            agg.unique_users_used.insert(h);
            agg.unique_user_hashes.insert(h);
        }
    }

    /// Record a feature viewed/rendered event.
    pub fn record_view(&self, feature: &str, identity: Option<&str>, variant: &str) {
        let hashed = identity.filter(|s| !s.is_empty()).map(hash_identity);
        let mut guard = self.inner.lock();
        if let Some(h) = hashed {
            guard.app_unique.insert(h);
        }
        let agg = guard.per_feature.entry(feature.to_string()).or_default();
        agg.variant_stats
            .entry(variant.to_string())
            .or_default()
            .viewed_count += 1;
        if let Some(h) = hashed {
            agg.unique_viewed_user_hashes.insert(h);
        }
    }

    /// True when no usage samples are buffered.
    pub fn is_empty(&self) -> bool {
        let guard = self.inner.lock();
        guard.per_feature.is_empty() && guard.app_unique.is_empty()
    }

    /// Build a FeatureStat-shaped payload and clear aggregates.
    ///
    /// Returns `(payload, snapshot)` so callers can restore unique maps on send failure.
    pub fn build_and_reset(&self) -> Option<(FeatureStatPayload, UsageSnapshot)> {
        let mut guard = self.inner.lock();
        if guard.per_feature.is_empty() && guard.app_unique.is_empty() {
            return None;
        }

        let snapshot = UsageSnapshot {
            per_feature: guard.per_feature.clone(),
            app_unique: guard.app_unique.clone(),
        };

        let mut stats_out = Vec::with_capacity(guard.per_feature.len());
        for (feature, agg) in guard.per_feature.drain() {
            let mut variant_stats = HashMap::new();
            for (name, vs) in agg.variant_stats {
                if !vs.is_empty() {
                    variant_stats.insert(name, vs.to_wire());
                }
            }
            stats_out.push(StatMessagePayload {
                feature,
                unique_context_identifier_enabled_count: agg.unique_users_enabled.len() as i32,
                unique_context_identifier_disabled_count: agg.unique_users_disabled.len() as i32,
                unique_users_used_count: agg.unique_users_used.len() as i32,
                unique_user_hashes: agg.unique_user_hashes.into_iter().collect(),
                unique_viewed_user_hashes: agg.unique_viewed_user_hashes.into_iter().collect(),
                variant_stats,
            });
        }

        let unique_user_hashes: Vec<i32> = guard.app_unique.drain().collect();
        let total_unique_users = unique_user_hashes.len() as i32;
        let process_start_time =
            to_protobuf_timestamp_millis(self.process_start_time.timestamp_millis());

        Some((
            FeatureStatPayload {
                app_key: self.app_key.clone(),
                environment: self.environment.clone(),
                time: now_protobuf_timestamp(),
                stats: stats_out,
                total_unique_users,
                unique_user_hashes,
                instance_name: self.instance_name.clone(),
                app_version: self.app_version.clone(),
                process_start_time,
            },
            snapshot,
        ))
    }

    /// Merge a previously drained snapshot back into the batcher after a failed send.
    ///
    /// Matches the .NET lesson: restore unique maps (and counters) so they are not lost.
    pub fn restore_snapshot(&self, snapshot: UsageSnapshot) {
        let mut guard = self.inner.lock();
        for hash in snapshot.app_unique {
            guard.app_unique.insert(hash);
        }
        for (feature, snap_agg) in snapshot.per_feature {
            let agg = guard.per_feature.entry(feature).or_default();
            for (name, vs) in snap_agg.variant_stats {
                let dest = agg.variant_stats.entry(name).or_default();
                dest.check_count = dest.check_count.saturating_add(vs.check_count);
                dest.request_count = dest.request_count.saturating_add(vs.request_count);
                dest.used_count = dest.used_count.saturating_add(vs.used_count);
                dest.viewed_count = dest.viewed_count.saturating_add(vs.viewed_count);
            }
            agg.unique_users_enabled
                .extend(snap_agg.unique_users_enabled);
            agg.unique_users_disabled
                .extend(snap_agg.unique_users_disabled);
            agg.unique_users_used.extend(snap_agg.unique_users_used);
            agg.unique_user_hashes.extend(snap_agg.unique_user_hashes);
            agg.unique_viewed_user_hashes
                .extend(snap_agg.unique_viewed_user_hashes);
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn aggregates_checks_into_variant_stats() {
        let batcher = UsageBatcher::new(
            "app",
            "Production",
            Some("host-1".into()),
            Some("1.2.3".into()),
            None,
        );
        batcher.record_check("FeatureA", true, Some("user-1"), None, false);
        batcher.record_check("FeatureA", true, Some("user-1"), None, false);
        batcher.record_check("FeatureA", false, Some("user-2"), None, false);
        batcher.record_usage("FeatureA", Some("user-1"), "enabled");
        batcher.record_view("FeatureA", Some("user-3"), "enabled");

        let (payload, _) = batcher.build_and_reset().expect("payload");
        assert_eq!(payload.app_key, "app");
        assert_eq!(payload.environment, "Production");
        assert_eq!(payload.instance_name.as_deref(), Some("host-1"));
        assert_eq!(payload.app_version.as_deref(), Some("1.2.3"));
        assert_eq!(payload.stats.len(), 1);

        let stat = &payload.stats[0];
        assert_eq!(stat.feature, "FeatureA");
        let enabled = stat.variant_stats.get("enabled").unwrap();
        assert_eq!(enabled.check_count, 2);
        assert_eq!(enabled.used_count, 1);
        assert_eq!(enabled.viewed_count, 1);
        let disabled = stat.variant_stats.get("disabled").unwrap();
        assert_eq!(disabled.check_count, 1);
        assert_eq!(stat.unique_context_identifier_enabled_count, 1);
        assert_eq!(stat.unique_context_identifier_disabled_count, 1);
        assert_eq!(stat.unique_users_used_count, 1);
        assert!(stat.unique_user_hashes.contains(&hash_identity("user-1")));
        assert!(stat
            .unique_viewed_user_hashes
            .contains(&hash_identity("user-3")));

        let mut app_hashes = payload.unique_user_hashes.clone();
        app_hashes.sort_unstable();
        let mut expected = vec![
            hash_identity("user-1"),
            hash_identity("user-2"),
            hash_identity("user-3"),
        ];
        expected.sort_unstable();
        assert_eq!(app_hashes, expected);
        assert!(batcher.build_and_reset().is_none());
    }

    #[test]
    fn restore_merges_unique_hashes_after_failed_send() {
        let batcher = UsageBatcher::new("app", "Production", None, None, None);
        batcher.record_check("FeatureA", true, Some("user-1"), None, false);
        batcher.record_usage("FeatureA", Some("user-1"), "enabled");
        let (_, snapshot) = batcher.build_and_reset().expect("payload");
        assert!(batcher.is_empty());

        batcher.restore_snapshot(snapshot);
        assert!(!batcher.is_empty());

        batcher.record_check("FeatureA", true, Some("user-2"), None, false);
        let (again, _) = batcher.build_and_reset().expect("merged");
        let mut hashes = again.unique_user_hashes.clone();
        hashes.sort_unstable();
        let mut expected = vec![hash_identity("user-1"), hash_identity("user-2")];
        expected.sort_unstable();
        assert_eq!(hashes, expected);
        let enabled = again.stats[0].variant_stats.get("enabled").unwrap();
        assert_eq!(enabled.check_count, 2);
        assert_eq!(enabled.used_count, 1);
        assert_eq!(again.stats[0].unique_context_identifier_enabled_count, 2);
    }

    #[test]
    fn request_count_only_when_unique_request() {
        let batcher = UsageBatcher::new("app", "Production", None, None, None);
        batcher.record_check("FeatureA", true, Some("user-1"), None, true);
        batcher.record_check("FeatureA", true, Some("user-1"), None, false);
        let (payload, _) = batcher.build_and_reset().unwrap();
        let enabled = payload.stats[0].variant_stats.get("enabled").unwrap();
        assert_eq!(enabled.check_count, 2);
        assert_eq!(enabled.request_count, 1);
    }
}
